server/src/service.rs

```diff
@@ -115,17 +115,61 @@ impl Service {
     #[zbus(out_args("unlocked", "prompt"))]
     pub async fn unlock(
         &mut self,
-        _objects: Vec<OwnedObjectPath>,
+        objects: Vec<OwnedObjectPath>,
     ) -> Result<(Vec<OwnedObjectPath>, ObjectPath), ServiceError> {
-        todo!()
+        let mut unlocked = Vec::new();
+        let prompt = ObjectPath::default();
+        let collections = self.collections.lock().await;
+
+        for object in &objects {
+            for collection in collections.iter() {
+                if object == collection.path() {
+                    if !collection.set_locked(false).await? {
+                        unlocked.push(object.clone());
+                    }
+                    tracing::info!("Collection: {} is locked.", object);
+                    break;
+                } else if let Some(item) = collection.item_from_path(object).await {
+                    if !item.set_locked(false).await? {
+                        unlocked.push(object.clone());
+                    }
+                    tracing::info!("Item: {} is locked.", object);
+                    break;
+                }
+                tracing::info!("Object: {} does not exist.", object);
+            }
+        }
+
+        Ok((unlocked, prompt))
     }
 
     #[zbus(out_args("locked", "prompt"))]
     pub async fn lock(
         &mut self,
-        _objects: Vec<OwnedObjectPath>,
+        objects: Vec<OwnedObjectPath>,
     ) -> Result<(Vec<OwnedObjectPath>, ObjectPath), ServiceError> {
-        todo!()
+        let mut locked = Vec::new();
```

Comment:
It is the exact same implementation, with only difference being the state of locked, put the common code in a function.

Reply:
Should I still follow this and move this to a new function?

Reply:
yes

Reply:
basically share what is common between both :)

Reply:
will do, thanks!